src/statsd_server.rs

```diff
@@ -182,55 +183,102 @@ async fn client_handler(
                 break;
             }
             Err(e) if e.kind() == ErrorKind::TimedOut => {
-                debug!("read timeout, closing {:?}", socket.peer_addr());
+                debug!("read timeout, closing {}", peer);
                 break;
             }
             Err(e) => {
-                debug!("socket error {:?} from {:?}", e, socket.peer_addr());
+                debug!("socket error {:?} {}", e, peer);
                 break;
             }
         }
     }
     disconnects.inc();
 }
 
-pub async fn run(stats: stats::Scope, tripwire: Tripwire, bind: String, backends: Backends) {
-    //self.shutdown_trigger = Some(trigger);
-    let listener = TcpListener::bind(bind.as_str()).await.unwrap();
+/// Wrapper type to adapt an optional listener and either return an accept
+/// future, or a pending future which never returns. This wrapper is needed to
+/// work around that .accept() is an opaque impl Future type, so can't be
+/// readily mixed into a stream.
+async fn optional_accept(
+    listener: Option<&UnixListener>,
+) -> std::io::Result<(UnixStream, unix::SocketAddr)> {
+    if listener.is_some() {
+        listener.unwrap().accept().await
+    } else {
+        futures::future::pending::<std::io::Result<(UnixStream, unix::SocketAddr)>>().await
```

Comment:
Can the parameter here be inferred from the function return type?

`future::pending().await` would be a little quicker to scan at a glance, if so